### PR TITLE
Implement gas estimation for EIP1559 transactions

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/18169

Easiest to review per commit.
Summary of commits:
- Add `GetEstimatedTime` and `GetGasOracle` in `AssetRatioController` to get time and gas estimation using etherscan API. (passthrough via bat-ratios server)
- Store `GasEstimation` info in Eip1559Transaction for UI to use.
- In `EthTxController::AddUnapproved1559Transaction`, 
    - Call `GetGasOracle` if max_priority_fee_per_gas or max_fee_per_gas is not specified, avg fees will be used by default.
    - Call `GetEstimateGas` if gas limit is not specified.
- Add `EthTxController::SetGasFeeAndLimitForUnapprovedTransaction` to update unapproved transaction.
- Expose to JS.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by kylehickinson — APPROVED** [review]
iOS++

File: components/brave_wallet/browser/eth_tx_controller.cc [hunk 1]
```diff
@@ -669,6 +669,51 @@ void EthTxController::SetGasPriceAndLimitForUnapprovedTransaction(
...
+  }
+
+  std::unique_ptr<EthTxStateManager::TxMeta> tx_meta =
+      tx_state_manager_->GetTx(tx_meta_id);
+  if (!tx_meta || tx_meta->status != mojom::TransactionStatus::Unapproved) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  auto* tx1559 = reinterpret_cast<Eip1559Transaction*>(tx_meta->tx.get());
```
Comment (darkdh) [inline components/brave_wallet/browser/eth_tx_controller.cc:691]: need to check `EthTransaction::Type() == 2` before continuing and we should also add a test case that trying to call this function for non eip1559 transaction and we should expect failure

File: components/brave_wallet/browser/eip1559_transaction_unittest.cc [hunk 2]
```diff
@@ -230,6 +262,67 @@ TEST(Eip1559TransactionUnitTest, FromTxData) {
...
+  EXPECT_EQ(tx->gas_estimation(), Eip1559Transaction::GasEstimation());
+
+  tx = Eip1559Transaction::FromTxData(
+      mojom::TxData1559::New(
+          mojom::TxData::New("0x01", "0x3E8", "0x989680",
+                             "0x3535353535353535353535353535353535353535",
+                             "0x2A", std::vector<uint8_t>{1}),
+          "0x15BE", "0x7B", "0x1C8", missing_fields_gas_estimation.Clone()),
+      false);
+  EXPECT_EQ(tx->gas_estimation(), Eip1559Transaction::GasEstimation());
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/eip1559_transaction_unittest.cc:285]: strict = true for the first one and strict = false for the second.

[the remark this answers is not in the document]
Comment (darkdh) [reply components/brave_wallet/browser/eip1559_transaction_unittest.cc:285]: gotcha

File: components/brave_wallet/browser/asset_ratio_response_parser.cc [hunk 3]
```diff
@@ -178,4 +180,180 @@ bool ParseAssetPriceHistory(
...
+
+  auto* result = payload->FindKey("result");
+  if (!result) {
+    return nullptr;
+  }
+
+  const base::DictionaryValue* result_dict;
+  if (!result->GetAsDictionary(&result_dict)) {
+    return nullptr;
+  }
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/asset_ratio_response_parser.cc]: Fixed in https://github.com/brave/brave-core/pull/10332/commits/cf0bca8aa04d9364bc79af15f9bb9343da8c565c

File: components/brave_wallet/browser/asset_ratio_response_parser.cc [hunk 4]
```diff
@@ -178,4 +180,180 @@ bool ParseAssetPriceHistory(
...
+
+  const base::DictionaryValue* payload_dict;
+  if (!payload->GetAsDictionary(&payload_dict)) {
+    return nullptr;
+  }
+
+  auto* result = payload->FindKey("result");
+  if (!result) {
+    return nullptr;
+  }
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/asset_ratio_response_parser.cc]: Fixed in cf0bca8

File: components/brave_wallet/browser/asset_ratio_response_parser.cc [hunk 5]
```diff
@@ -178,4 +180,180 @@ bool ParseAssetPriceHistory(
...
+
+  auto* payload = response_dict->FindKey("payload");
+  if (!payload) {
+    return nullptr;
+  }
+
+  const base::DictionaryValue* payload_dict;
+  if (!payload->GetAsDictionary(&payload_dict)) {
+    return nullptr;
+  }
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/asset_ratio_response_parser.cc]: Fixed in cf0bca8

File: components/brave_wallet/browser/asset_ratio_response_parser.cc [hunk 6]
```diff
@@ -178,4 +180,180 @@ bool ParseAssetPriceHistory(
...
+
+  const base::DictionaryValue* payload_dict;
+  if (!payload->GetAsDictionary(&payload_dict)) {
+    return "";
+  }
+
+  const std::string* result = payload->FindStringKey("result");
+  if (!result) {
+    return "";
+  }
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/asset_ratio_response_parser.cc]: Fixed in cf0bca8

File: components/brave_wallet/browser/eth_tx_controller.cc (hunk 1 above)
Comment (yrliou) [reply components/brave_wallet/browser/eth_tx_controller.cc:692]: Good call, fixed in https://github.com/brave/brave-core/pull/10332/commits/df3df47c0a637144c7ff37269037e5b3bdfd01f0.

**Review by darkdh — APPROVED** [review]

**Pull request merged**